pymtl3/datatypes/bit_structs.py

```diff
@@ -0,0 +1,442 @@
+                    add_hash=True ):
+
+  fields = {}
+  add_eq = True
+
+  # Get annotations of the class
+  cls_annotations = cls.__dict__.get('__annotations__', {})
+  if not cls_annotations:
+    raise NoFieldDeclaredError(
+      'No field is delacred in the bit struct definition' )
+
+  # Get field information from the annotation
+  cls_fields = [ _get_field( cls, a_name, a_type )
+                 for a_name, a_type in cls_annotations.items() ]
+
+  # Create a dictionary of fields
+  for f in cls_fields:
+    fields[ f.name ] = f
+
+  # Stamp the special attribute so that translation pass can identify it
+  # as bit struct.
+  setattr( cls, _IS_BIT_STRUCT, fields )
+
+  # Sanity check: is there any Field instance that doesn't have an
+  # annotation?
+  for name, value in cls.__dict__.items():
+    if isinstance( value, Field ) and not name in cls_annotations:
+      raise TypeError(f'{name} is a field but has no type annotation!')
+
+  # Add methods to the class
+
+  # Create __init__. Here I follow the dataclass convention that we only
+  # add our generated __init__ function when add_init is true and user
+  # did not define their own init.
+  if add_init:
+    if not '__init__' in cls.__dict__:
+      self_name = _get_self_name( fields )
+      cls.__init__ = _mk_init_fn( self_name, fields.values() )
+
+  # Create __str
+  if add_str:
+    if not '__str__' in cls.__dict__:
+      cls.__str__ = _mk_str_fn( fields.values() )
+
+  # Create __repr__
+  if add_repr:
+    if not '__repr__' in cls.__dict__:
+      cls.__repr__ = _mk_repr_fn( fields.values() )
+
+  # Create __eq__. There is no need for a __ne__ method as python will
+  # call __eq__ and negate it.
+  # NOTE: if user overwrites __eq__ it may lead to different behavior for
+  # the translated verilog as in the verilog world two bit structs are
+  # equal only if all the fields are equal.
+  if add_eq:
+    if not '__eq__' in cls.__dict__:
+      cls.__eq__ = _mk_eq_fn( fields.values() )
+    else:
+      w_msg = ( f'Overwriting {cls.__qualname__}\'s __eq__ may cause the '
+                'translated verilog behaves differently from PyMTL '
+                'simulation.')
+      warnings.warn( w_msg )
+
+  # Create __hash__.
+  if add_hash:
+    if not '__hash__' in cls.__dict__:
+      cls.__hash__ = _mk_hash_fn( fields.values() )
+
+  # TODO: maybe add a to_bits and from bits function.
+
+  return cls
+
+#-------------------------------------------------------------------------
+# bit_struct
+#-------------------------------------------------------------------------
+# The actual class decorator. We add a * in the argument list so that the
+# following argument can only be used as keyword arguments.
+
+def bit_struct( _cls=None, *, add_init=True, add_str=True, add_repr=True,
+                add_hash=True ):
+
+  def wrap( cls ):
+    return _process_class( cls, add_init, add_str, add_repr )
+
+  # Called as @bitstruct(...)
```

Comment:
what does this do?

Reply:
This is to check if the decorator has any argument. We can use the decorator like
~~~
@bit_struct
class A:
  ...
~~~
or 
~~~
@bit_struct( add_repr = False )
class A:
  ...
~~~

Reply:
Hmm this new argument seperator might be able to solve some problem I encountered when I was playing around with method port decorator.